---
title: "Capstone 2 Prediction of ALL vs AML from Gene Expression"
author: "Dr. Hans M. Rupp"
date: "29 7 2022"
output: pdf_document
---


# Introduction

The purpose of this project is to evaluate different machine learning algorithms for predicting two different cancer types, Acute Myeloid Leukemia (AML) and Acute lymphoblastic leukaemia (ALL) from microarray RNA expression data.
AML and ALL are the two main types of Acute Leukemias. They differ in their origin and treatments. To provide optimal treatment the cancer type must be diagnosed [1]. Microarray Expression chips are arrays of oligonucleotides which hybridize to mRNAs of specific genes. They allow the semiquantitative analysis of the expression a great number of genes in parallel. One common application is the comparision of diseased vs disease free or of different diseases to find differentially expressed (DE) genes [2]

I will use the AML ALL microarray dataset by Golub [3]. The csv files can be downloaded from
https://www.kaggle.com/datasets/crawford/gene-expression

```{r message=FALSE, warning=FALSE }
rm(list=ls())
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
if (!require("limma"))
  install("limma")
if (!require("Biobase"))
  install("Biobase")

library(tidyverse)
library(caret)
library(matrixStats)
library(limma)
library(Biobase)


set.seed(2, sample.kind = "Rounding")
```


# Loading the data
```{r message=FALSE,warning=FALSE}
train_data <- read_csv("data_set_ALL_AML_train.csv", col_names = T)
test_data <- read_csv("data_set_ALL_AML_independent.csv")
labels <- read_csv("actual.csv")

```
# Checking and cleaning the data

Both train_data and test_data contain call columns which will not be used. We are only interested in gene expression data.
```{r}
head(train_data)
```
```{r}
train_data <- train_data %>% select(-contains("call"))
test_data <- test_data %>% select(-contains("call"))
```
Checking for missing values
```{r}
nas <- function(x) ( any(is.na(x)) )

print(train_data %>% summarise_all(nas), width = Inf)
print(test_data %>% summarise_all(nas), width = Inf)
```
There are no missing values.

We well merge the train_data and test_data set. Training and Testing data for ML alogorithms will be separated further down

```{r}
all_data <- train_data %>% left_join(test_data)
```

Inspecting the all_data set with View(all_data) shows that the data columns are not sorted according to the patient number.
Sorting:

```{r}
all_data_desc <- all_data[, 1:2]
all_data_data <- all_data[, 3:ncol(all_data)]
all_data_data <- all_data_data[, order(as.numeric(names(all_data_data)))]

all_data <- cbind(all_data_desc, all_data_data)
```

all_data now contains two columns with descriptions of the genes analyzed and `r ncol(all_data) -2` columns with gene expression data from the different patients. `r nrow(all_data)` genes are in rows.
labels contains  `r nrow(labels)` entries with patient number and cancer type.

We have `r nrow(all_data)` gene expression levels. It is highly unlikely that all or most of them are differentially regulated between these related tumors. Most will only contribute noise to any analysis. In a first step it is necessary to identify the differentially expressed (DE) genes.

# Selection of differentially  expressed genes

DE genes could be identified with a Student's T-test. However the package limma provides a more robust and convenient implementation for DE identification. [4] Limma is part of the Bioconductor project at http://www.bioconductor.org.

Construction of a Biobase ExpressionSet Object
```{r}
#features
f <- all_data[, 1:2]

#expression matrix
x <- as.matrix(all_data[, 3:ncol(all_data)])


#Colnames for expression set necessary
colnames(x) <- 1:ncol(x)

eset <- ExpressionSet(assayData = x,
                      phenoData = AnnotatedDataFrame(labels),
                      featureData = AnnotatedDataFrame(f))
```

Mean-reference model [5]

ALL and AML are the factors or classifiers associated with the samples in the experiment.

Y - Expression level of the gene

$\beta{_1}$ - Expression level for ALL

$\beta{_2}$ - Difference of expression between AML and ALL

Y = $\beta{_1}$ + $\beta{_2}$AML


Design Matrix

Design matrices are used in the estimation process of model parameters. The design matrix has columns associated with the parameters and rows associated with samples [5]
```{r}
design <- model.matrix(~cancer, data=pData(eset))
```
The number of rows in the design matrix is the number of observations
```{r}
nrow(design)
```
The number of AML observations is equal to the number of AML observations in the label set
```{r}
colSums(design)
sum(labels$cancer=="AML")
```

Estimate the fold changes and standard errors by tting a linear model for each gene

```{r}

fit <- lmFit(eset, design)
```
Apply empirical Bayes smoothing to the standard errors

```{r}
fit <- eBayes(fit)

results <- decideTests(fit[, "cancerAML"])
summary(results)
```

Top 100 differentially expressed genes
```{r}
top_DE_genes <- topTable(fit, n=100, adjust="fdr")
head(top_DE_genes)
```

Get the data of the top 100 DE genes
```{r}
top_data <- all_data %>% filter(`Gene Accession Number` %in% top_DE_genes$Gene.Accession.Number)
```


Comparison of the expression of the top DE gene
```{r}
zyxin <- top_data %>% filter(`Gene Description`=="Zyxin") %>% select(-c(1,2)) 
zyxin <- as.numeric(t(zyxin))
zyxindf <- data.frame(x = zyxin, labels)
zyxindf %>% ggplot(aes(x=cancer, y=x)) + geom_boxplot() + labs(title = "Zyxin Expression", x="Cancer", y="Expression")
```




# References
[1] Demystifying Targeted Cancer Treatment. Course. Cancer Research UK
[2] Lesk, Arthur M. : Introduction to Genomics. Oxford 2017.
[3] Golub et. al.: Molecular Classification of Cancer: Class Discovery and Class Prediction by Gene Expression Monitoring. Science 286, 531 (1999)
[4] Ritchie, ME, Phipson, B, Wu, D, Hu, Y, Law, CW, Shi, W, and Smyth, GK (2015).
limma powers differential expression analyses for RNA-sequencing and microarray studies.
Nucleic Acids Research 43(7), e47.
[5] https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html